### PR TITLE
`relational.write` block: operation to be determined based on `opcode_field` #191

### LINKED ISSUE #191: `relational.write` block: operation to be determined based on $$opcode field
If specified, assume that this field exists in the data and contains:
c - create
d- delete
u - update

It also means that this determines the load strategy, so the load strategy is not requested if the `opcode_field` is passed.

### PR DESCRIPTION
**Pull request by spicy-sauce**
upsert support for postgresql

### CODE REVIEW

File: core/src/datayoga_core/blocks/relational/write/block.py [hunk 1]
```diff
@@ -10,14 +12,48 @@
...
+        db_type: str) -> Any:
+    if db_type.lower() == "postgresql":
+        update_fields = [f"{field['column']} = {sa.bindparam(field['key'])}" for field in mapping_fields]
+
+        insert_fields = ", ".join([field["column"] for field in mapping_fields])
+        pk_fields = ", ".join([field["column"] for field in primary_keys])
+
+        insert_bind_params = ", ".join([f"{sa.bindparam(field['key'])}" for field in mapping_fields])
+
+        return sa.text(f"""INSERT INTO {table} ({insert_fields})
```
Comment (zalmane) [inline core/src/datayoga_core/blocks/relational/write/block.py:40]: there is an sql builder for the upsert now in SA. https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#insert-on-conflict-upsert.
we should move to the API in next iteration as we add more dialects.

File: core/src/datayoga_core/blocks/relational/write/block.py [hunk 2]
```diff
@@ -27,17 +63,67 @@ def init(self, context: Optional[Context] = None):
...
+            # group records by opcode
+            opcodes = [{"opcode": key, "records": list(result)} for key, result in groupby(
+                data, key=lambda record: record.get(self.opcode_field, "").replace(OpCode.CREATE.value, OpCode.UPDATE.value))]
+
+            for records_by_opcode in opcodes:
+                opcode = records_by_opcode["opcode"]
+                records = records_by_opcode["records"]
+
+                if opcode == OpCode.UPDATE.value:
+                    logger.info(f"Upserting {len(records)} record(s) to {self.table} table")
```
Comment (zalmane) [inline core/src/datayoga_core/blocks/relational/write/block.py]: change to debug

Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/relational/write/block.py]: done

File: core/src/datayoga_core/blocks/relational/write/block.py (hunk 1 above)
Comment (spicy-sauce) [reply core/src/datayoga_core/blocks/relational/write/block.py:40]: I'll open an issue for that.

**Review by zalmane — APPROVED** [review]

**Pull request merged**